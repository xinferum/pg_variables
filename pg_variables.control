# pg_variables extension
comment = 'session variables with various types'
default_version = '1.2'
module_pathname = '$libdir/pg_variables'
relocatable = true
