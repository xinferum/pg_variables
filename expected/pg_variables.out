CREATE EXTENSION pg_variables;
-- Test packages - sanity checks
SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

SELECT pgv_exists(NULL); -- fail
ERROR:  package name can not be NULL
SELECT pgv_remove(NULL); -- fail
ERROR:  package name can not be NULL
SELECT pgv_remove('vars'); -- fail
ERROR:  unrecognized package "vars"
SELECT pgv_exists('vars111111111111111111111111111111111111111111111111111111111111'); -- fail
ERROR:  name "vars111111111111111111111111111111111111111111111111111111111111" is too long
-- Integer variables
SELECT pgv_get_int('vars', 'int1');
ERROR:  unrecognized package "vars"
SELECT pgv_get_int('vars', 'int1', false);
 pgv_get_int 
-------------
            
(1 row)

SELECT pgv_set_int('vars', 'int1', 101);
 pgv_set_int 
-------------
 
(1 row)

SELECT pgv_set_int('vars', 'int2', 102);
 pgv_set_int 
-------------
 
(1 row)

SELECT pgv_get_int('vars', 'int1');
 pgv_get_int 
-------------
         101
(1 row)

SELECT pgv_get_int('vars', 'int2');
 pgv_get_int 
-------------
         102
(1 row)

SELECT pgv_set_int('vars', 'int1', 103);
 pgv_set_int 
-------------
 
(1 row)

SELECT pgv_get_int('vars', 'int1');
 pgv_get_int 
-------------
         103
(1 row)

SELECT pgv_get_int('vars', 'int3');
ERROR:  unrecognized variable "int3"
SELECT pgv_get_int('vars', 'int3', false);
 pgv_get_int 
-------------
            
(1 row)

SELECT pgv_exists('vars', 'int3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'int1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_exists('vars2');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_set_int('vars', 'intNULL', NULL);
 pgv_set_int 
-------------
 
(1 row)

SELECT pgv_get_int('vars', 'intNULL');
 pgv_get_int 
-------------
            
(1 row)

-- Text variables
SELECT pgv_set_text('vars', 'str1', 's101');
 pgv_set_text 
--------------
 
(1 row)

SELECT pgv_set_text('vars', 'int1', 's101');
ERROR:  variable "int1" requires "integer" value
SELECT pgv_set_int('vars', 'str1', 101);
ERROR:  variable "str1" requires "text" value
SELECT pgv_set_text('vars', 'str2', 's102');
 pgv_set_text 
--------------
 
(1 row)

SELECT pgv_get_text('vars', 'str1');
 pgv_get_text 
--------------
 s101
(1 row)

SELECT pgv_get_text('vars', 'str2');
 pgv_get_text 
--------------
 s102
(1 row)

SELECT pgv_set_text('vars', 'str1', 's103');
 pgv_set_text 
--------------
 
(1 row)

SELECT pgv_get_text('vars', 'str1');
 pgv_get_text 
--------------
 s103
(1 row)

SELECT pgv_get_text('vars', 'str3');
ERROR:  unrecognized variable "str3"
SELECT pgv_get_text('vars', 'str3', false);
 pgv_get_text 
--------------
 
(1 row)

SELECT pgv_exists('vars', 'str3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'str1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_text('vars', 'int1');
ERROR:  variable "int1" requires "integer" value
SELECT pgv_get_int('vars', 'str1');
ERROR:  variable "str1" requires "text" value
SELECT pgv_set_text('vars', 'strNULL', NULL);
 pgv_set_text 
--------------
 
(1 row)

SELECT pgv_get_text('vars', 'strNULL');
 pgv_get_text 
--------------
 
(1 row)

-- Numeric variables
SELECT pgv_set_numeric('vars', 'num1', 1.01);
 pgv_set_numeric 
-----------------
 
(1 row)

SELECT pgv_set_numeric('vars', 'num2', 1.02);
 pgv_set_numeric 
-----------------
 
(1 row)

SELECT pgv_set_numeric('vars', 'str1', 1.01);
ERROR:  variable "str1" requires "text" value
SELECT pgv_get_numeric('vars', 'num1');
 pgv_get_numeric 
-----------------
            1.01
(1 row)

SELECT pgv_get_numeric('vars', 'num2');
 pgv_get_numeric 
-----------------
            1.02
(1 row)

SELECT pgv_set_numeric('vars', 'num1', 1.03);
 pgv_set_numeric 
-----------------
 
(1 row)

SELECT pgv_get_numeric('vars', 'num1');
 pgv_get_numeric 
-----------------
            1.03
(1 row)

SELECT pgv_get_numeric('vars', 'num3');
ERROR:  unrecognized variable "num3"
SELECT pgv_get_numeric('vars', 'num3', false);
 pgv_get_numeric 
-----------------
                
(1 row)

SELECT pgv_exists('vars', 'num3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'num1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_numeric('vars', 'str1');
ERROR:  variable "str1" requires "text" value
SELECT pgv_set_numeric('vars', 'numNULL', NULL);
 pgv_set_numeric 
-----------------
 
(1 row)

SELECT pgv_get_numeric('vars', 'numNULL');
 pgv_get_numeric 
-----------------
                
(1 row)

SET timezone = 'Europe/Moscow';
-- Timestamp variables
SELECT pgv_set_timestamp('vars', 'ts1', '2016-03-30 10:00:00');
 pgv_set_timestamp 
-------------------
 
(1 row)

SELECT pgv_set_timestamp('vars', 'ts2', '2016-03-30 11:00:00');
 pgv_set_timestamp 
-------------------
 
(1 row)

SELECT pgv_set_timestamp('vars', 'num1', '2016-03-30 12:00:00');
ERROR:  variable "num1" requires "numeric" value
SELECT pgv_get_timestamp('vars', 'ts1');
    pgv_get_timestamp     
--------------------------
 Wed Mar 30 10:00:00 2016
(1 row)

SELECT pgv_get_timestamp('vars', 'ts2');
    pgv_get_timestamp     
--------------------------
 Wed Mar 30 11:00:00 2016
(1 row)

SELECT pgv_set_timestamp('vars', 'ts1', '2016-03-30 12:00:00');
 pgv_set_timestamp 
-------------------
 
(1 row)

SELECT pgv_get_timestamp('vars', 'ts1');
    pgv_get_timestamp     
--------------------------
 Wed Mar 30 12:00:00 2016
(1 row)

SELECT pgv_get_timestamp('vars', 'ts3');
ERROR:  unrecognized variable "ts3"
SELECT pgv_get_timestamp('vars', 'ts3', false);
 pgv_get_timestamp 
-------------------
 
(1 row)

SELECT pgv_exists('vars', 'ts3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'ts1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_timestamp('vars', 'num1');
ERROR:  variable "num1" requires "numeric" value
SELECT pgv_set_timestamp('vars', 'tsNULL', NULL);
 pgv_set_timestamp 
-------------------
 
(1 row)

SELECT pgv_get_timestamp('vars', 'tsNULL');
 pgv_get_timestamp 
-------------------
 
(1 row)

-- TimestampTZ variables
SELECT pgv_set_timestamptz('vars', 'tstz1', '2016-03-30 10:00:00 GMT+01');
 pgv_set_timestamptz 
---------------------
 
(1 row)

SELECT pgv_set_timestamptz('vars', 'tstz2', '2016-03-30 11:00:00 GMT+02');
 pgv_set_timestamptz 
---------------------
 
(1 row)

SELECT pgv_set_timestamptz('vars', 'ts1', '2016-03-30 12:00:00 GMT+03');
ERROR:  variable "ts1" requires "timestamp without time zone" value
SELECT pgv_get_timestamptz('vars', 'tstz1');
     pgv_get_timestamptz      
------------------------------
 Wed Mar 30 14:00:00 2016 MSK
(1 row)

SELECT pgv_get_timestamptz('vars', 'tstz2');
     pgv_get_timestamptz      
------------------------------
 Wed Mar 30 16:00:00 2016 MSK
(1 row)

SELECT pgv_set_timestamptz('vars', 'tstz1', '2016-03-30 12:00:00 GMT+01');
 pgv_set_timestamptz 
---------------------
 
(1 row)

SELECT pgv_get_timestamptz('vars', 'tstz1');
     pgv_get_timestamptz      
------------------------------
 Wed Mar 30 16:00:00 2016 MSK
(1 row)

SELECT pgv_get_timestamptz('vars', 'tstz3');
ERROR:  unrecognized variable "tstz3"
SELECT pgv_get_timestamptz('vars', 'tstz3', false);
 pgv_get_timestamptz 
---------------------
 
(1 row)

SELECT pgv_exists('vars', 'tstz3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'tstz1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_timestamptz('vars', 'ts1');
ERROR:  variable "ts1" requires "timestamp without time zone" value
SELECT pgv_set_timestamptz('vars', 'tstzNULL', NULL);
 pgv_set_timestamptz 
---------------------
 
(1 row)

SELECT pgv_get_timestamptz('vars', 'tstzNULL');
 pgv_get_timestamptz 
---------------------
 
(1 row)

-- Date variables
SELECT pgv_set_date('vars', 'd1', '2016-03-29');
 pgv_set_date 
--------------
 
(1 row)

SELECT pgv_set_date('vars', 'd2', '2016-03-30');
 pgv_set_date 
--------------
 
(1 row)

SELECT pgv_set_date('vars', 'tstz1', '2016-04-01');
ERROR:  variable "tstz1" requires "timestamp with time zone" value
SELECT pgv_get_date('vars', 'd1');
 pgv_get_date 
--------------
 03-29-2016
(1 row)

SELECT pgv_get_date('vars', 'd2');
 pgv_get_date 
--------------
 03-30-2016
(1 row)

SELECT pgv_set_date('vars', 'd1', '2016-04-02');
 pgv_set_date 
--------------
 
(1 row)

SELECT pgv_get_date('vars', 'd1');
 pgv_get_date 
--------------
 04-02-2016
(1 row)

SELECT pgv_get_date('vars', 'd3');
ERROR:  unrecognized variable "d3"
SELECT pgv_get_date('vars', 'd3', false);
 pgv_get_date 
--------------
 
(1 row)

SELECT pgv_exists('vars', 'd3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars', 'd1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_date('vars', 'tstz1');
ERROR:  variable "tstz1" requires "timestamp with time zone" value
SELECT pgv_set_date('vars', 'dNULL', NULL);
 pgv_set_date 
--------------
 
(1 row)

SELECT pgv_get_date('vars', 'dNULL');
 pgv_get_date 
--------------
 
(1 row)

-- Jsonb variables
SELECT pgv_set_jsonb('vars2', 'j1', '[1, 2, "foo", null]');
 pgv_set_jsonb 
---------------
 
(1 row)

SELECT pgv_set_jsonb('vars2', 'j2', '{"bar": "baz", "balance": 7.77, "active": false}');
 pgv_set_jsonb 
---------------
 
(1 row)

SELECT pgv_set_jsonb('vars', 'd1', '[1, 2, "foo", null]');
ERROR:  variable "d1" requires "date" value
SELECT pgv_get_jsonb('vars2', 'j1');
    pgv_get_jsonb    
---------------------
 [1, 2, "foo", null]
(1 row)

SELECT pgv_get_jsonb('vars2', 'j2');
                  pgv_get_jsonb                   
--------------------------------------------------
 {"bar": "baz", "active": false, "balance": 7.77}
(1 row)

SELECT pgv_set_jsonb('vars2', 'j1', '{"foo": [true, "bar"], "tags": {"a": 1, "b": null}}');
 pgv_set_jsonb 
---------------
 
(1 row)

SELECT pgv_get_jsonb('vars2', 'j1');
                    pgv_get_jsonb                    
-----------------------------------------------------
 {"foo": [true, "bar"], "tags": {"a": 1, "b": null}}
(1 row)

SELECT pgv_get_jsonb('vars2', 'j3');
ERROR:  unrecognized variable "j3"
SELECT pgv_get_jsonb('vars2', 'j3', false);
 pgv_get_jsonb 
---------------
 
(1 row)

SELECT pgv_exists('vars2', 'j3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars2', 'j1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_get_jsonb('vars', 'd1');
ERROR:  variable "d1" requires "date" value
SELECT pgv_set_jsonb('vars', 'jNULL', NULL);
 pgv_set_jsonb 
---------------
 
(1 row)

SELECT pgv_get_jsonb('vars', 'jNULL');
 pgv_get_jsonb 
---------------
 
(1 row)

-- Record variables
CREATE TABLE tab (id int, t varchar);
INSERT INTO tab VALUES (0, 'str00'), (1, 'str11'), (2, NULL), (NULL, 'strNULL');
SELECT pgv_insert('vars3', 'r1', tab) FROM tab;
 pgv_insert 
------------
 
 
 
 
(4 rows)

SELECT pgv_insert('vars2', 'j1', tab) FROM tab;
ERROR:  variable "j1" requires "jsonb" value
SELECT pgv_insert('vars3', 'r1', tab) FROM tab;
ERROR:  there is a record in the variable "r1" with same key
SELECT pgv_insert('vars3', 'r1', row(1, 'str1', 'str2'));
ERROR:  new record structure have 3 attributes, but variable "r1" structure have 2.
SELECT pgv_insert('vars3', 'r1', row(1, 1));
ERROR:  new record attribute type for attribute number 2 differs from variable "r1" structure.
HINT:  You may need explicit type casts.
SELECT pgv_insert('vars3', 'r1', row('str1', 'str1'));
ERROR:  new record attribute type for attribute number 1 differs from variable "r1" structure.
HINT:  You may need explicit type casts.
SELECT pgv_select('vars3', 'r1', ARRAY[[1,2]]); -- fail
ERROR:  searching for elements in multidimensional arrays is not supported
-- Test variables caching
SELECT pgv_insert('vars3', 'r2', row(1, 'str1', 'str2'));
 pgv_insert 
------------
 
(1 row)

SELECT pgv_update('vars3', 'r1', row(3, 'str22'::varchar));
 pgv_update 
------------
 f
(1 row)

SELECT pgv_update('vars4', 'r1', row(3, 'str22'::varchar)); -- fail
ERROR:  unrecognized package "vars4"
select pgv_delete('vars3', 'r2', NULL::int);
 pgv_delete 
------------
 f
(1 row)

select pgv_delete('vars4', 'r2', NULL::int); -- fail
ERROR:  unrecognized package "vars4"
-- Test NULL values
SELECT pgv_insert('vars3', 'r2', NULL); -- fail
ERROR:  record argument can not be NULL
SELECT pgv_update('vars3', 'r2', NULL); -- fail
ERROR:  record argument can not be NULL
select pgv_delete('vars3', 'r2', NULL::int);
 pgv_delete 
------------
 f
(1 row)

SELECT pgv_select('vars3', 'r1', NULL::int[]); -- fail
ERROR:  array argument can not be NULL
SELECT pgv_select('vars3', 'r1');
 pgv_select 
------------
 (,strNULL)
 (1,str11)
 (2,)
 (0,str00)
(4 rows)

SELECT pgv_select('vars3', 'r1', 1);
 pgv_select 
------------
 (1,str11)
(1 row)

SELECT pgv_select('vars3', 'r1', 1::float); -- fail
ERROR:  requested value type differs from variable "r1" key type
SELECT pgv_select('vars3', 'r1', 0);
 pgv_select 
------------
 (0,str00)
(1 row)

SELECT pgv_select('vars3', 'r1', NULL::int);
 pgv_select 
------------
 (,strNULL)
(1 row)

SELECT pgv_select('vars3', 'r1', ARRAY[1, 0, NULL]);
 pgv_select 
------------
 (1,str11)
 (0,str00)
 (,strNULL)
(3 rows)

UPDATE tab SET t = 'str33' WHERE id = 1;
SELECT pgv_update('vars3', 'r1', tab) FROM tab;
 pgv_update 
------------
 t
 t
 t
 t
(4 rows)

SELECT pgv_update('vars3', 'r1', row(4, 'str44'::varchar));
 pgv_update 
------------
 f
(1 row)

SELECT pgv_select('vars3', 'r1');
 pgv_select 
------------
 (,strNULL)
 (1,str33)
 (2,)
 (0,str00)
(4 rows)

SELECT pgv_delete('vars3', 'r1', 1);
 pgv_delete 
------------
 t
(1 row)

SELECT pgv_select('vars3', 'r1');
 pgv_select 
------------
 (,strNULL)
 (2,)
 (0,str00)
(3 rows)

SELECT pgv_delete('vars3', 'r1', 100);
 pgv_delete 
------------
 f
(1 row)

SELECT pgv_select('vars3', 'r3');
ERROR:  unrecognized variable "r3"
SELECT pgv_exists('vars3', 'r3');
 pgv_exists 
------------
 f
(1 row)

SELECT pgv_exists('vars3', 'r1');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_select('vars2', 'j1');
ERROR:  variable "j1" requires "jsonb" value
-- PGPRO-2601 - Test pgv_select() on TupleDesc of dropped table
DROP TABLE tab;
SELECT pgv_select('vars3', 'r1');
 pgv_select 
------------
 (,strNULL)
 (2,)
 (0,str00)
(3 rows)

-- Tests for SRF's sequential scan of an internal hash table
DO
$$BEGIN
    PERFORM pgv_select('vars3', 'r1') LIMIT 2 OFFSET 2;
    PERFORM pgv_select('vars3', 'r3');
END$$;
ERROR:  unrecognized variable "r3"
CONTEXT:  SQL statement "SELECT pgv_select('vars3', 'r3')"
PL/pgSQL function inline_code_block line 3 at PERFORM
-- Check that the hash table was cleaned up after rollback
SET client_min_messages to 'ERROR';
SELECT pgv_select('vars3', 'r1', 1);
 pgv_select 
------------
 
(1 row)

SELECT pgv_select('vars3', 'r1') LIMIT 2; -- warning
 pgv_select 
------------
 (,strNULL)
 (2,)
(2 rows)

SELECT pgv_select('vars3', 'r1') LIMIT 2 OFFSET 2;
 pgv_select 
------------
 (0,str00)
(1 row)

-- PGPRO-2601 - Test a cursor with the hash table
BEGIN;
DECLARE r1_cur CURSOR FOR SELECT pgv_select('vars3', 'r1');
FETCH 1 in r1_cur;
 pgv_select 
------------
 (,strNULL)
(1 row)

SELECT pgv_select('vars3', 'r1');
 pgv_select 
------------
 (,strNULL)
 (2,)
 (0,str00)
(3 rows)

FETCH 1 in r1_cur;
 pgv_select 
------------
 (2,)
(1 row)

CLOSE r1_cur;
COMMIT; -- warning
RESET client_min_messages;
-- Clean memory after unsuccessful creation of a variable
SELECT pgv_insert('vars4', 'r1', row (('str1'::text, 'str1'::text))); -- fail
ERROR:  could not identify a hash function for type record
SELECT package FROM pgv_stats() WHERE package = 'vars4';
 package 
---------
(0 rows)

-- Remove package if it is empty
SELECT pgv_insert('vars4', 'r2', row(1, 'str1', 'str2'));
 pgv_insert 
------------
 
(1 row)

SELECT pgv_remove('vars4', 'r2');
 pgv_remove 
------------
 
(1 row)

SELECT package FROM pgv_stats() WHERE package = 'vars4';
 package 
---------
(0 rows)

-- Record variables as scalar
SELECT pgv_set('vars5', 'r1', row(1, 'str11'));
 pgv_set 
---------
 
(1 row)

SELECT pgv_get('vars5', 'r1', NULL::record);
  pgv_get  
-----------
 (1,str11)
(1 row)

SELECT pgv_set('vars5', 'r1', row(1, 'str11'), true); -- fail
ERROR:  variable "r1" already created as NOT TRANSACTIONAL
SELECT pgv_insert('vars5', 'r1', row(1, 'str11')); -- fail
ERROR:  "r1" isn't a record variable
SELECT pgv_select('vars5', 'r1'); -- fail
ERROR:  "r1" isn't a record variable
SELECT pgv_get('vars3', 'r1', NULL::record); -- fail
ERROR:  "r1" isn't a scalar variable
-- Manipulate variables
SELECT * FROM pgv_list() order by package, name;
 package |   name   | is_transactional 
---------+----------+------------------
 vars    | d1       | f
 vars    | d2       | f
 vars    | dNULL    | f
 vars    | int1     | f
 vars    | int2     | f
 vars    | intNULL  | f
 vars    | jNULL    | f
 vars    | num1     | f
 vars    | num2     | f
 vars    | numNULL  | f
 vars    | str1     | f
 vars    | str2     | f
 vars    | strNULL  | f
 vars    | ts1      | f
 vars    | ts2      | f
 vars    | tsNULL   | f
 vars    | tstz1    | f
 vars    | tstz2    | f
 vars    | tstzNULL | f
 vars2   | j1       | f
 vars2   | j2       | f
 vars3   | r1       | f
 vars3   | r2       | f
 vars5   | r1       | f
(24 rows)

SELECT package FROM pgv_stats() order by package;
 package 
---------
 vars
 vars2
 vars3
 vars5
(4 rows)

SELECT pgv_remove('vars', 'int3');
ERROR:  unrecognized variable "int3"
SELECT pgv_remove('vars', 'int1');
 pgv_remove 
------------
 
(1 row)

SELECT pgv_get_int('vars', 'int1');
ERROR:  unrecognized variable "int1"
SELECT pgv_exists('vars');
 pgv_exists 
------------
 t
(1 row)

SELECT pgv_remove('vars2');
 pgv_remove 
------------
 
(1 row)

SELECT pgv_get_jsonb('vars2', 'j1');
ERROR:  unrecognized package "vars2"
SELECT pgv_exists('vars2');
 pgv_exists 
------------
 f
(1 row)

SELECT * FROM pgv_list() order by package, name;
 package |   name   | is_transactional 
---------+----------+------------------
 vars    | d1       | f
 vars    | d2       | f
 vars    | dNULL    | f
 vars    | int2     | f
 vars    | intNULL  | f
 vars    | jNULL    | f
 vars    | num1     | f
 vars    | num2     | f
 vars    | numNULL  | f
 vars    | str1     | f
 vars    | str2     | f
 vars    | strNULL  | f
 vars    | ts1      | f
 vars    | ts2      | f
 vars    | tsNULL   | f
 vars    | tstz1    | f
 vars    | tstz2    | f
 vars    | tstzNULL | f
 vars3   | r1       | f
 vars3   | r2       | f
 vars5   | r1       | f
(21 rows)

SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

SELECT pgv_exists('vars');
 pgv_exists 
------------
 f
(1 row)

SELECT * FROM pgv_list() order by package, name;
 package | name | is_transactional 
---------+------+------------------
(0 rows)

-- Check insert of record with various amount of fields
CREATE TEMP TABLE foo(id int, t text);
INSERT INTO foo VALUES (0, 'str00');
SELECT pgv_insert('vars', 'r1', row(1, 'str1'::text, 'str2'::text));
 pgv_insert 
------------
 
(1 row)

SELECT pgv_select('vars', 'r1');
  pgv_select   
---------------
 (1,str1,str2)
(1 row)

SELECT pgv_insert('vars', 'r1', foo) FROM foo;
ERROR:  new record structure have 2 attributes, but variable "r1" structure have 3.
SELECT pgv_select('vars', 'r1');
  pgv_select   
---------------
 (1,str1,str2)
(1 row)

SELECT pgv_insert('vars', 'r2', row(1, 'str1')); -- ok, UNKNOWNOID of 'str1' converts to TEXTOID
 pgv_insert 
------------
 
(1 row)

SELECT pgv_insert('vars', 'r2', foo) FROM foo; -- ok
 pgv_insert 
------------
 
(1 row)

SELECT pgv_select('vars', 'r2');
 pgv_select 
------------
 (1,str1)
 (0,str00)
(2 rows)

SELECT pgv_insert('vars', 'r3', row(1, 'str1'::text));
 pgv_insert 
------------
 
(1 row)

SELECT pgv_insert('vars', 'r3', foo) FROM foo; -- ok, no conversions
 pgv_insert 
------------
 
(1 row)

SELECT pgv_select('vars', 'r3');
 pgv_select 
------------
 (1,str1)
 (0,str00)
(2 rows)

SELECT pgv_insert('vars', 'r4', row(1, 2::int));
 pgv_insert 
------------
 
(1 row)

SELECT pgv_insert('vars', 'r4', row(0, 'str1')); -- fail, UNKNOWNOID of 'str1' can't be converted to int
ERROR:  new record attribute type for attribute number 2 differs from variable "r4" structure.
HINT:  You may need explicit type casts.
SELECT pgv_select('vars', 'r4');
 pgv_select 
------------
 (1,2)
(1 row)

SELECT pgv_insert('vars', 'r5', foo) FROM foo; -- types: int, text
 pgv_insert 
------------
 
(1 row)

SELECT pgv_insert('vars', 'r5', row(1, 'str1')); -- ok, UNKNOWNOID of 'str1' converts to TEXTOID
 pgv_insert 
------------
 
(1 row)

SELECT pgv_select('vars', 'r5');
 pgv_select 
------------
 (1,str1)
 (0,str00)
(2 rows)

